src/upgrade/upgrade_scripts/5.12.0/upgrade_objectmds_old_versions.js

```diff
@@ -0,0 +1,86 @@
+/* Copyright (C) 2021 NooBaa */
+"use strict";
+
+const util = require('util');
+
+async function run({ dbg, system_store, md_store }) {
+
+    try {
+        if (process.env.DB_TYPE !== 'postgres') {
+            dbg.log0('upgrade postgres objectmds latest index: not running on postgres - no upgrade needed...');
+            return;
+        }
+        dbg.log0('starting upgrade version past on non latest object versions...');
+
+        // The following upgrade process will do the following:
+        // 1. update old versions of an object with version_past = true
+        // 2. drop the current index - idx_btree_objectmds_latest_version_index
+
+        // idx_btree_objectmds_latest_version_index postgres index is incorrect since 
+        // postgres considers nulls as distinct, causes multiple latest versions of objectmds to be allowed
+        // recreation of the index is being handled lazily in md_store constructor
+
+        // 1. update old versions
+        const versioned_buckets = system_store.data.buckets.filter(bucket => bucket.versioning === 'ENABLED');
+
+        if (versioned_buckets.length > 0) {
+            dbg.log0(`upgrade_objectmds: upgrading version past buckets: 
+                ${versioned_buckets.map(bucket => util.inspect(bucket)).join(', ')}`);
+
+            // update entries as non latest
+            const update = `UPDATE objectmds SET data = jsonb_set(data, '{version_past}', '"true"') `;
+            // self join of obectmds
+            const select_fields = `SELECT distinct A._id FROM objectmds A, objectmds B `;
+            // select lower version_seq entry when the 2 entries matche by key and bucket
+            // and the 2 entries are marked as latest version of the object (version_past = null & delete = null)
+            const select_by = `A.data->>'key' = B.data->>'key' 
+                AND ((A.data -> 'version_past'::text) IS NULL OR (A.data -> 'version_past'::text) = 'null'::jsonb) 
+                AND ((B.data -> 'version_past'::text) IS NULL OR (B.data -> 'version_past'::text) = 'null'::jsonb)
+                AND ((A.data -> 'deleted'::text) IS NULL OR (A.data -> 'deleted'::text) = 'null'::jsonb) 
+                AND ((B.data -> 'deleted'::text) IS NULL OR (B.data -> 'deleted'::text) = 'null'::jsonb)
+                AND ((A.data -> 'upload_started'::text) IS NULL OR (A.data -> 'upload_started'::text) = 'null'::jsonb)
+                AND ((B.data -> 'upload_started'::text) IS NULL OR (B.data -> 'upload_started'::text) = 'null'::jsonb) 
+                AND A.data->'version_seq' < B.data->'version_seq'`;
+
+            for (let bucket of versioned_buckets) {
+                try {
+                    dbg.log0('upgrade_objectmds: updating bucket', bucket._id);
+                    const query = `${update} WHERE _id in (${select_fields} WHERE
+                        A.data->>'bucket'='${bucket._id}' AND B.data->>'bucket'='${bucket._id}'
+                        AND ${select_by});`;
```

Comment:
I don't think you need to run over the versioned buckets - you repeat the scan per bucket.
but if you just match on the A.data->>'bucket' = B.data->>'bucket' like you do for keys it will be a single query for all the buckets.

Reply:
```
nbcore=# EXPLAIN
SELECT A._id
FROM objectmds A, objectmds B
WHERE A.data->>'bucket' = B.data->>'bucket'
AND A.data->>'key' = B.data->>'key'
AND ((A.data -> 'version_past'::text) IS NULL OR (A.data -> 'version_past'::text) = 'null'::jsonb)
AND ((B.data -> 'version_past'::text) IS NULL OR (B.data -> 'version_past'::text) = 'null'::jsonb)
AND ((A.data -> 'deleted'::text) IS NULL OR (A.data -> 'deleted'::text) = 'null'::jsonb)
AND ((B.data -> 'deleted'::text) IS NULL OR (B.data -> 'deleted'::text) = 'null'::jsonb)
AND ((A.data -> 'upload_started'::text) IS NULL OR (A.data -> 'upload_started'::text) = 'null'::jsonb)
AND ((B.data -> 'upload_started'::text) IS NULL OR (B.data -> 'upload_started'::text) = 'null'::jsonb)
AND A.data->>'version_seq' < B.data->>'version_seq'
;
                                                                                                    QUERY PLAN
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Nested Loop  (cost=0.28..16.34 rows=1 width=25)
   Join Filter: (((a.data ->> 'version_seq'::text) < (b.data ->> 'version_seq'::text)) AND ((a.data ->> 'bucket'::text) = (b.data ->> 'bucket'::text)) AND ((a.data ->> 'key'::text) = (b.data ->> 'key'::text)))
   ->  Index Scan using idx_btree_objectmds_latest_version_index on objectmds a  (cost=0.14..8.15 rows=1 width=508)
   ->  Index Scan using idx_btree_objectmds_latest_version_index on objectmds b  (cost=0.14..8.15 rows=1 width=483)
(4 rows)

```

Reply:
yes but I'm afraid it'll take so much time so we will see a session timeout, so I thought maybe we should still iterate over the buckets. WDYT?

Reply:
Can you set a custom timeout for this connection?

Reply:
I can try,
@tamireran and I are currently doing some performance tests on a 200K objects dataset and only for one bucket we can see it takes 13 seconds.
currently, I'm running it for all buckets and it didn't finish even after more than 10 minutes.

Reply:
Well but even a single bucket can contain all the objects... so we need to be prepared for a longer call. Did you try the query I suggested in the pr review or the one in the code now?

Reply:
@guymguym I tried both

Reply:
Can you try just the select without the update? Is that significantly shorter?

Reply:
ok I found the issue, there was missing > in the left null check

Reply:
and now what? is it fast?

Reply:
the select of all buckets now takes 10 seconds, which is good.
The update is very slow

Reply:
->> - returns output but query plan doesn't contain index scan
-> - doesn't return output but query plan contains an index scan

Reply:
Afaiu data->>'key' is equivalent to data->'key'::text

Reply:
https://www.postgresql.org/docs/current/functions-json.html

Reply:
Can you show what is your final suggestion? I'm a bit lost

Reply:
```
UPDATE objectmds SET data = jsonb_set(data, '{version_past}', '"true"') 
WHERE _id in (SELECT distinct A._id FROM objectmds A, objectmds B
WHERE (A.data->>'bucket'=B.data->>'bucket')
AND (A.data->>'key' = B.data->>'key')
AND ((A.data ->> 'version_past'::text) IS NULL OR (A.data -> 'version_past'::text) = 'null'::jsonb)
AND ((B.data ->> 'version_past'::text) IS NULL OR (B.data -> 'version_past'::text) = 'null'::jsonb)
AND ((A.data ->> 'deleted'::text) IS NULL OR (A.data -> 'deleted'::text) = 'null'::jsonb)
AND ((B.data ->> 'deleted'::text) IS NULL OR (B.data -> 'deleted'::text) = 'null'::jsonb)
AND ((A.data ->> 'upload_started'::text) IS NULL OR (A.data -> 'upload_started'::text) = 'null'::jsonb)
AND ((B.data ->> 'upload_started'::text) IS NULL OR (B.data -> 'upload_started'::text) = 'null'::jsonb)
AND (A.data->'version_seq' < B.data->'version_seq'));
```

this query runs forever and even when specifing bucket id it doesn't work for large dataset

Reply:
resolving since we fixed it together.